### ISSUE TITLE
Document that StreamBuilder renders no-data frame even when Stream has a value right away, and give alternatives

### ISSUE THREAD

**Issue by alanrussian** [flutter/flutter#16465]
Internal: b/140829693

We frequently have a need for a `StreamBuilder` type widget in our app (mulligan) but there's a major issue with it that prevents us from using it: if the stream has an initial value when the widget is first built, the widget will still be built with a no data frame before the widget is built with a data frame. This results in jank that we cannot have in our app.

The sequence of events seems to be the following today:

1. `StreamBuilder` constructor is called.
2. `StreamBuilder` state is created. The current snapshot is set to the loading snapshot.
3. `StreamBuilder` state's `initState` method is called. Stream is listened to.
4. `StreamBuilder` state's `build` method is called with the loading snapshot.
5. `StreamBuilder` state's stream listener is invoked. `setState` is called with the data snapshot and the widget is rebuilt.

As far as I understand, the problem is that when you listen to a `Stream` in Dart, it doesn't fire the callback synchronously. Because of this, I'm not sure if there's anything that can be done to resolve this issue. That said, until this gets fixed, I think that `StreamBuilder` is unusable in this use case. To start discussion, does Flutter perform the steps from 2-4 in the same run-loop?

Here's a test that depicts the problem I'm discussing: https://gist.github.com/alanrussian/d4a0d8c779275832dad44d4acde9cc59

**Comment by jonahwilliams** [comment]
@alanrussian  In the case where you've already loaded the data, why doesn't [`initialData`](https://github.com/flutter/flutter/blob/b494e71e365ffa48d2562276756919f7aca9fd56/packages/flutter/lib/src/widgets/async.dart#L374) work?

**Comment by alanrussian** [comment]
The already loaded data is in the stream, so we don't have direct access to it in order to pass it into `initialData`. The typical use case here is that we have some model exposed as a Stream that already has data but we want to support updating the widget on subsequent updates.

**Comment by newfivefour** [comment]
Could you take the last element of your stream and pass that to `initialData`?

**Comment by alanrussian** [comment]
That would work if we had the last element of the stream available; however, if you're given just a `Stream` and don't have its last value, there's no way to get the last value from the `Stream` object when constructing the `StreamBuilder`. My knowledge of Dart streams is also not the best, though, so maybe there's something else we could do.

**Comment by jonahwilliams** [comment]
@alanrussian  So there are (roughly) four different kinds of Streams, not counting any custom implementations.

 * sync brodcast
 * sync single subscription
 * async broadcast
 * async single subscription

It's not possible for a broadcast stream to have an initial state.  Either you were subscribed when the data was added or you missed it, so those are out.  In an async single-subscription stream, any listen calls added won't be invoked until either the next microtask or next event loop (can't remember, may depend), but at any rate there is no way to get the data out the stream on the current frame. In the case of a sync single subscription stream, the StreamBuilder works as you would expect - see https://gist.github.com/jonahwilliams/e8d71584c54d9c7c2d8dffb21f70d061.

It's possible you have logic wired up in the controller, so that `onListen` adds some cached data to the stream?  That falls into the same category as above.



Any way, if yo

**Comment by alanrussian** [comment]
I tried using a sync stream and was unable to get it to work. If you add `sync: true` to the `StreamController` constructor in the test in my OP, the test fails. I added "assert(false);" in your example to where you expect the text to never be displayed, and the assertion was did fail.

From a discussion with @lrhn, the issue here is that, "if you have a synchronous controller, you *can* deliver events immediately to *already subscribed* listeners, but a new listener isn't considered subscribed until the listen call returns the new subscription."

**Comment by jonahwilliams** [comment]
ahh you're right, I was mistaken.  It looks like even in the case of a sync single subscription stream the subscription callback fires sometime after the first build.

So another approach that might be adjusting the architecture to expose an update stream and current state, or using a custom stream implementation that has a cached `currentState` member.  That way your could pass it to `initialState` while still getting stream updates?

**Comment by alanrussian** [comment]
We currently have what's effectively a custom stream implementation
(technically, it's an RX implementation) that makes this a non-issue for
us; however, one of our long term goals is to move off of it in favor of
dart streams.

It would be great if flutter could somehow solve this problem with dart
streams as they exist (or by proposing modifications to the dart streams),
which is why I opened the issue. We don't need a short term fix. A fix
would be helpful for us and presumably others in the long term.

**Comment by jonahwilliams** [comment]
Is it important that it is a Stream, or would another interface make sense?  For example, Flutter has a [`ValueNotifier`](https://github.com/flutter/flutter/blob/b9a7f1b915349210c7a80b1ac158ed43fd88f612/packages/flutter/lib/src/foundation/change_notifier.dart#L203) class which provides both a current value and a way of getting updates.  You could conceivably use this or a similar class and write a `ValueListenerBuilder` widget which would avoid any FOUC-style rebuilds when data is cached. 

For a new implementation of Stream, you easily write a Stream + Controller pair which had mostly synchronous behavior.  Changes to the Stream implementation is more challenging, since it would require changing the behavior across Flutter, VM, and web.

**Comment by newfivefour** [comment]
@alanrussian Out of interest, are you using StreamBuilder with your custom stream implementation?

**Comment by alanrussian** [comment]
Jonah, it is definitely desirable for us to move toward the stream API. All
the oeprators it comes with make it really useful. Perhaps if an
implementation change were made to it, it could be behind a flag? I'm also
curious about whether there's anything Flutter can do by changing
StreamBuilder's implementation but I'm guessing not.

NewFiveFour, yes we are. Since our custom stream allows us to get the
current value and it allows us to convery it to a Stream class, it's pretty
easy to hook it up to StreamBuilder.

**Comment by Hixie** [comment]
cc @mravn-google

**Comment by Hixie** [comment]
cc @floitschG

**Comment by mravn-google** [comment]
There is nothing much that `StreamBuilder` can do about the fact that the first item in the `Stream` is not synchronously available. Seems to me you don't really have a Dart `Stream` as your data type, but rather something like

```dart
class HeadedStream<T> {
  final T first;
  final Stream<T> rest;
  ...
}
```

`StreamBuilder` should be useful with that:
```dart
StreamBuilder(initialData: headedStream.first, stream: headedStream.rest, builder: ...);
```

**Comment by alanrussian** [comment]
Sorry for the delayed response. I was on vacation.

I agree that there's nothing much `StreamBuilder` can do today. I'm raising this issue because:
- I believe this issue is a significant drawback to using `StreamBuilder` in apps (again, not in Flutter's control, but a drawback nonetheless). 
- I want to start discussion ASAP to see if any long term fixes (e.g., Dart API changes) are possible.

We have to go to great lengths to avoid using `StreamBuilder`. We use an entirely different stream-like (RX) framework to get around issues such as this one.

@mravn-google the issue with creating something like `HeadedStream` is it doesn't work well with stream operators. While the example you provided is easy to handle, it gets trickier if you have to compose multiple `HeadedStream`s and apply operators like map. You end up having to duplicate code in a non-trivial way.

**Comment by jonahwilliams** [comment]
@alanrussian I would open up an issue on the SDK repo and outline some of the cases where Stream behavior is either not intuitive or doesn't quite do what you need.  I don't really know much about the design decisions of the Stream class, so I can't really say if these changes are realistic or not.

**Comment by alanrussian** [comment]
I think there's a very similar issue with `Future`s. Let's say I have an API to get some data, and I want it to fetch the data on the first call but return cached data immediately on subsequent calls. With Dart, it's unclear to me how to do this because a future always completes in the next runloop (see `Future.value`). I suppose I could return FutureOr but that seems like a weird return type. 

Will post this to the dart issue as well.

**Comment by Hixie** [comment]
Flutter gets around this with [SynchronousFuture](https://master-docs-flutter-io.firebaseapp.com/flutter/foundation/SynchronousFuture-class.html), but it's a bit sketchy. We use this for image loading, for instance, where image loads of cached images must be synchronous.

**Comment by alanrussian** [comment]
Cool, thanks, Ian. It would be really nice to get better APIs for these issues in Dart. Especially for our case since we share code across web and mobile.

**Comment by rrousselGit** [comment]
> We currently have what's effectively a custom stream implementation
(technically, it's an RX implementation) that makes this a non-issue for
us


@alanrussian Would this be possible to share it with the community? This is quite a big issue with BLoC pattern being more and more popular. 

I can work on one if it's impossible to share, but yours is probably more complete.

**Comment by alanrussian** [comment]
I think you could use the already published rx dart package

**Comment by figengungor** [comment]
Hi @alanrussian Can you have a look at the SO question here? How do you use rxdart in this scenario?

**Comment by rrousselGit** [comment]
RxDart is simply a wrapper of StreamController, so it suffers from the same issue: The first event is asynchronous. 

There are some way to hack around it, by altering the purity of Streams (exposing their current value, and using that as `initialData` on`StreamBuilder`). But that's miles away from a clean architecture

**Comment by alanrussian** [comment]
I haven't tried this myself but the `BehaviorSubject` class in RxDart exposes a getter with the last value that I thought you could pass as the `initialData`.

https://pub.dartlang.org/documentation/rxdart/latest/rx/BehaviorSubject/value.html

**Comment by barrylapthorn** [comment]
I see the same problem, but it is due to the fact that I mixed `setState` calls and `StreamBuilder` in the same `Widget`.  That has the effect of the `setState` causing a re-`build` of the whole visual tree and the `StreamBuilder` being recreated time after time and missing the initial `Stream` event and leaving the `StreamBuilder` in `ConnectionState.waiting`

In my case there also seems to be a correlation between the `Stream` in question being a very fast ticking one versus slow ticking ones.

`BehaviorSubject` didn't work for me.

In summary, what worked for me was to have a `Widget` that used exclusively `Stream`s, or `setState` but not both.

**Comment by rrousselGit** [comment]
> I haven't tried this myself but the `BehaviorSubject` class in RxDart exposes a getter with the last value that I thought you could pass as the `initialData`.

Using `initialData` this way is more of a hack than anything.
It is very common to hide a `BehaviorSubject` under the `Stream` interface, which makes using `initialData` impossible when it theorically shouldn't change anything.

**Comment by DanT29** [comment]
I have a similar problem that I posted [here](https://stackoverflow.com/questions/54761678/stream-is-not-re-rendering-when-switching-tabs-on-flutter/54777093?noredirect=1#comment96511598_54777093).  I had two tabs each with a `StreamBuilder` one had data the other didn't. The data is acquired via a GET request from a server this was done with the BLoC architecture.  When I was on the tab without data and then navigated away to another page and came back to the tab that had data it would **sometimes** show there was no data. My streams are all `BehaviourSubject`. Should I be using `FutureBuilder` instead?

**Comment by modulovalue** [comment]
I'd like to share my solution to this problem.

Currently, in my opinion, the StreamBuilder along with pure dart streams is completely 'useless'.
Business logic and presentation should always be separated. Every example on the web gives the UI the responsibility of providing the initialData.

```dart
/// Random example
return new StreamBuilder<PlayerState>(
    stream: _streamController.stream,
    initialData: new PlayerState.initial(),
    builder: (BuildContext context, AsyncSnapshot<PlayerState> snapshot) {
      final playerState = snapshot.data;
      return buildUi(context, playerState);
    });
```

This is terrible and the UI should not be concerned with what the initial state is.

This is currently best solved with a BehaviorSubject (RxDart). A BehaviorSubject has a synchronously accessible initialData (value) and is also a ValueObservable with which the sink can be hidden from the presentation.

Now, one can construct a new 'StreamBuilder'.

```dart
class ValueObservableBuilder<T> extends StatelessWidget {
  final ValueObservable<T> stream;
  final Widget Function(BuildContext, T) builder;

  const ValueObservableBuilder({
    @required this.stream,
    @required this.builder,
  });

  @override
  Widget build(BuildContext context) {
    return StreamBuilder<T>(
        stream: stream,
        initialData: stream.value,
        builder: (BuildContext context, AsyncSnapshot<T> snapshot) {
          if (snapshot.hasError) {
            // ignore: only_throw_errors
            throw snapshot.error; // Handle as you wish
          }
          return builder(context, snapshot.data);
        });
  }
}
``` 

Now, all would be great if there wasn't a problem with ValueObservables.

ValueObservables cannot be transformed into ValueObservables. Every method on every Observable transforms into an Observable. This isn't a problem for counters and hello-world  examples, but a real-world application might want to transform observables many layers deep to prepare that data for the presentation. And if the source Stream has synchronously available data then every transformation is also technically able to derive its initial value from the source stream.

My solution to this problem is the following:


```dart
/// Tested
    static ValueObservable<T> where<T>(ValueObservable<T> stream, bool Function(T) test) {
        return stream.where(test).skip(1).shareValueSeeded(stream.value);
    }

    static ValueObservable<S> switchMap<A, S>(ValueObservable<A> observable, ValueObservable<S> Function(A) mapper) {
        return observable.switchMap(mapper).skip(1).shareValueSeeded(mapper(observable.value).value);
    }

    static ValueObservable<R> map1<A, R>(ValueObservable<A> observable, R Function(A) map) {
        return observable.map(map).skip(1).shareValueSeeded(map(observable.value));
    }
...
```
This will become way more usable as soon as static extension methods are implemented. 

Now, There is only one problem left that @rrousselGit has gracefully solved with his flutter_hooks package

Take a look at this 
```dart
return ValueObservableBuilder<Widget>(
        stream: base().api.name,
        builder: (_, __) {
          return ValueObservableBuilder<Widget>(
              stream: base().api.baz,
              builder: (_, __) {
                return ValueObservableBuilder<Widget>(
                    stream: base().api.count,
                    builder: (_, __) {
                      return ValueObservableBuilder<Widget>(
                          stream: base().api.foo,
                          builder: (_, __) {
                            return ValueObservableBuilder<Widget>(
                                stream: base().api.bar,
                                builder: (_, child) {
                                  return child;
                                });
                            ;
                          });
                      ;
                    });
                ;
              });
        });
```

this is terrible but can easily be simplified with the following hook:

```dart
/// Tested, be careful, you might need to memoize the ValueObservable
T useVO<T>(ValueObservable<T> Function() valueObservable) {
    final vo = valueObservable();
    final a = useStream<T>(vo, initialData: vo.value, );
    if (a.hasError) {
        // ignore: only_throw_errors
        throw a.error; /// Handle as you wish
    } else {
        return a.data;
    }
}
```

And now one can do the following:

```dart
Widget build(BuildContext context) {
  final name = useVO(() => base().api.name);
  final foo = useVO(() => base().api.name);
  final bar = useVO(() => base().api.name);
  final baz = useVO(() => base().api.name);
  final anything = useVO(() => base().api.name);
  return buildSomething(name, foo, baz, bar, anything);
}
```

My solution has a steep learning curve (Rx! and Hooks) but does solve this issue very well. I'm currently in the midst of testing every step of it, so take it with a grain of salt.

**Comment by Hixie** [comment]
In general, `ValueListenable`s are the solution to this problem. They are like lightweight streams. They work well with `ValueListenableBuilder`s.

**Comment by modulovalue** [comment]
@Hixie 
`ValueListenable`s are tied to flutter. And that is the problem for those who'd like to share their code between flutter and non-flutter projects (which is one of the main selling points of flutter?).
Even if flutter should take over every platform there is, there will still be use-cases where flutter cannot be used as a dependency (e.g. server applications) and that rules out `ValueListenable` as _the_ solution to this problem.

**Comment by rrousselGit** [comment]
> In general, ValueListenables are the solution to this problem.

Agreed. But through my adventure on the different mediums used by the Flutter community, it seems like peoples refuse to accept `Listenable` as a replacement to `Stream`.

A few of the arguments I've seen are:

- `Listenable` are just reinventing streams.
- Streams are simple enough.
- Things like Firebase uses streams, so out of consistency, we should use them everywhere.

I also noted that there are multiple attempts from the community at pushing `Listenable`. But they are usually received negatively because RxDart is just _that_ popular.

If `Listenable` is the accepted solution to the multiple issues faced by streams, then it may be interesting for Google to make an official statement on that matter.

**Comment by Hixie** [comment]
> `ValueListenable`s are tied to flutter

They're very simple and under a BSD license, feel free to copy the code into your project so you can use it on the server.

> it may be interesting for Google to make an official statement on that matter.

I'm not sure how much more official it can get that me (I'm Flutter's TL) saying that "In general, `ValueListenable`s are the solution to this problem".

You are welcome to use whatever features you like. Streams will always have the problem described in this issue, so long as they don't expose their "current value" synchronously. At some level, one has to make a decision: does one want to use streams, or does one want to avoid having the widget show one frame with no data.

Incidentally, it's relatively straightforward to adapt between from listenables to streams and futures, if the issue is that you have code that expects streams and so you don't want to use listenables. See, for example, this code: https://github.com/flutter/flutter/compare/master...Hixie:listenable_stream

**Comment by modulovalue** [comment]
> They're very simple and under a BSD license, feel free to copy the code into your project so you can use it on the server.

But that misses the whole point. @Hixie please let me elaborate on why ValueListenables are not a satisfactory replacement for streams.

Let's say we have the following mono-repo structure

```
my_crossplatform_app
 |- business_logic
 |- flutter
 |- angular
 |- backend_service
```

The goal here is to share as much code as possible between flutter/angular and the backend_service. The business_logic should obviously not depend on flutter, angular or any other framework prescribed by the UI (ValueListenables).

The [DarfConf 2018 BLoC presentation](https://www.youtube.com/watch?v=PLHln7wHgPE) goes into greater detail about the point I'm trying to make here and presents Bloc's as a solution but the reason why @alanrussian opened this issue is why Bloc's are so terrible to use. And if one would really try to replace streams with ValueListenables then:

1. one really had to copy that code and write wrappers for angular and for flutter to use his own copy of ValueListenables with flutters own ValueListenables
or
2. one would tightly bind his business logic to flutter and it's hopefully obvious why that is not a great solution.

--

I don't mean to sound rude and please forgive if I am but are you aware of the underlying issue here?

React has Redux.
Android has Android Architecture Components.
Flutter has ...?
The problem is that the community wants (expects) a little guidance in that regard. Everybody expected a go-to solution during the last IO and now provider [is the official solution for state management](https://flutter.dev/docs/development/data-and-backend/state-mgmt/simple)?

Quote from [Android Architecture Components](https://developer.android.com/topic/libraries/architecture)
> Android architecture components are a collection of libraries that help you design *robust*, *testable*, and *maintainable apps*. (emphasis mine)

How can we design robust, testable and maintainable flutter apps (an additional requirement should be so that one is able to share code between flutter & other UI frameworks like Angular Dart)

Shouldn't you as the TL have the ability to go and talk to the Angular & Dart team and find a real solution? You mentioned streams with a synchronously available value, yes, please! that would indeed be a solution.

PS:
For the past 16 months, I've been literally putting every hour that I have into Dart & Flutter. I truly believe that flutter will overtake everything and @Hixie I can't thank you enough for the work so far.
But there's no official architecture suitable for apps that go beyond a basic hello world. 
Why is that important? This would lead the way for a whole new ecosystem of tutorials, courses, trust (and with that developers), and issues like this one wouldn't happen.

**Comment by rrousselGit** [comment]
@modulovalue if you think that the architecture presented during the IO is not scalable (which, unless I'm misunderstanding you're message, is what you're suggesting) then please make an issue on `provider`'s repository with your concerns.

**Comment by Hixie** [comment]
I mean at the end of the day, how you manage your state really is up to you. We recommend @rrousselGit's `provider` package as an excellent starting point, but if you think about it, "managing state" is basically a pretty accurate description of "programming" in general. There will never be a single solution that solves everyone's problems.

That's somewhat out of scope for this issue, though, which is about a limitation of `StreamBuilder`. I haven't closed this issue because we need to document the limitation better.

**Comment by derolf** [comment]
I am having the same issue and am using ValueListenables now. Is there some package that offers functional operators for VLs (map, combineLatest, ...)?

**Comment by vladostaci** [comment]
I've tried using this workaround for now. Any ideas how to make this use simpler until a better solution ?
```
StreamBuilder<String>(
  initialData: _bloc.initialData(_bloc.inputText),
  stream: _bloc.inputText,
  builder: (context, snapshot) {

}
```
In BLoC I have:
```
final _inputTextSubject = BehaviorSubject.seeded(_initialInputText);
Stream<String> get inputText => _inputTextSubject.stream;

initialData(Stream stream) {
  if (stream == _inputTextSubject) {
    return _initialInputText;
  }
}
```

**Comment by jkurtw** [comment]
b/140829693

**Comment by Comevius** [comment]
@vladostaci

```
ValueObservable<String> get inputText => _inputTextSubject.stream;
```
```
StreamBuilder<String> {
  initialData: _bloc.inputText.value,
  stream: _bloc.inputText,
  builder: (context, snapshot) {}
}
```

**Comment by markusaksli-nc** [comment]
Closing as this has now been documented in [initialData](https://api.flutter.dev/flutter/widgets/StreamBuilder/initialData.html). Follow up on https://github.com/dart-lang/sdk/issues/30532 for more on this functionality.
